// @framer/krijn.video-player/code/Video.tsx
import * as React from "react"
import { PropertyControls, ControlType, RenderTarget } from "framer"

enum SrcType {
    Video = "Video",
    Url = "URL",
}

interface Props {
    width: number
    height: number
    srcType: SrcType
    video: string
    url: string
    autoplay: boolean
    loop: boolean
    muted: boolean
    controls: boolean
}

export class VideoPlayer extends React.Component<Props> {
    static defaultProps = {
        width: 320,
        height: 180,
        srcType: SrcType.Url,
        url: "https://assets.mixkit.co/videos/1489/1489-720.mp4",
        video: "",
        autoplay: true,
        loop: true,
        muted: true,
        controls: true,
    }

    static propertyControls: PropertyControls<Props> = {
        srcType: {
            type: ControlType.SegmentedEnum,
            title: "Source",
            options: [SrcType.Url, SrcType.Video],
        },
        url: {
            type: ControlType.String,
            title: " ",
            placeholder: ".../example.mp4",
            hidden(props) {
                return props.srcType === SrcType.Video
            },
        },
        video: {
            type: ControlType.File,
            title: " ",
            allowedFileTypes: ["mp4"],
            hidden(props) {
                return props.srcType === SrcType.Url
            },
        },
        autoplay: { type: ControlType.Boolean, title: "Play" },
        loop: { type: ControlType.Boolean, title: "Loop" },
        muted: { type: ControlType.Boolean, title: "Muted" },
        controls: { type: ControlType.Boolean, title: "Controls" },
    }

    render() {
        let {
            width,
            height,
            srcType,
            url,
            video,
            autoplay,
            loop,
            muted,
            controls,
        } = this.props

        if (RenderTarget.current() != RenderTarget.preview) {
            controls = false
            muted = true
        }

        return (
            <div
                style={{
                    width: "100%",
                    height: "100%",
                    overflow: "hidden",
                    backgroundColor: "#000",
                    color: "#333",
                    fontSize: "14px",
                }}
            >
                <video
                    key={`#${autoplay}`}
                    autoPlay={autoplay}
                    loop={loop}
                    muted={muted}
                    controls={controls}
                    preload="auto"
                    playsInline
                    style={{
                        width: "100%",
                        height: "100%",
                        objectFit: "cover",
                        objectPosition: "50% 50%",
                    }}
                    src={srcType === SrcType.Url ? url : video}
                />
                Your browser does not support HTML5 video.
            </div>
        )
    }
}
